src/hotspot/share/opto/mulnode.cpp

```diff
@@ -594,36 +594,61 @@ const Type* MulHiValue(const Type *t1, const Type *t2, const Type *bot) {
   return TypeLong::LONG;
 }
 
+template<typename IntegerType>
+static const IntegerType* and_value(const IntegerType* r0, const IntegerType* r1) {
+  typedef typename IntegerType::NativeType NativeType;
+
+  int widen = MAX2(r0->_widen,r1->_widen);
+
+  // If both are constants, we can calculate a precise result.
+  if(r0->is_con() && r1->is_con()) {
+    return IntegerType::make(r0->get_con() & r1->get_con());
+  }
+
+  // If both ranges are positive, the result will range from 0 up to the maximum value of the smaller range.
+  if (r0->_lo >= 0 && r1->_lo >= 0) {
+    return IntegerType::make(0, MIN2(r0->_hi, r1->_hi), widen);
+  }
+
+  // If only one range is positive, the result will range from 0 up to that range's maximum value.
+  if (r0->_lo >= 0) {
+    return IntegerType::make(0, r0->_hi, widen);
+  }
+
+  if (r1->_lo >= 0) {
+    return IntegerType::make(0, r1->_hi, widen);
+  }
+
+  // At this point, all positive ranges will have already been handled, so the only remaining cases will be negative ranges
+  // and constants.
+
+  assert(r0->_lo < 0 && r1->_lo < 0, "positive ranges should already be handled!");
+  static_assert(std::is_signed<NativeType>::value, "native type of IntegerType must be signed!");
+
+  // The lower bound of both ranges will contain the common leading 1-bits. In order to count them with count_leading_zeros,
+  // the bits are inverted.
+  NativeType sel_val = ~MIN2(r0->_lo, r1->_lo);
+
+  // To get the number of bits to shift, we count the leading 0-bits and then subtract one, as the sign bit is already set.
+  // Since count_leading_zeros is undefined at 0 (~(-1)) the number of digits in the native type can be used instead,
+  // as it returns 31 and 63 for signed integers and longs respectively.
+  int shift_bits = sel_val == 0 ? std::numeric_limits<NativeType>::digits : count_leading_zeros(sel_val) - 1;
```

Comment:
`sel_val` can only be 0 if `r0->_lo` and `r1->_lo` are both -1. While I think it's correct how you handle the case here, wouldn't it be simpler/more readable if you handle this case separately by setting -1 as lower bound directly instead of using "`min >> #digits"`?

Reply:
This is a good point! The logic becomes a lot easier to parse when we don't need to work around `count_leading_zeros` being undefined at 0 and just set the lower bound manually.